### PR TITLE
Add admin api authentication function

### PR DESCRIPTION
**Pull request by KazuhiroItoh**
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Proposed changes
This PR proposes a feature that enables the execution of the Admin API even in the 'strategy' scheme by adding a customizable token-based authentication function.

Details are written in [design note](https://github.com/node-red/designs/blob/master/designs/admin-api-authentication.md).

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [x] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

File: packages/node_modules/@node-red/editor-api/lib/auth/index.js [hunk 1]
```diff
@@ -60,7 +61,7 @@ function init(_settings,storage) {
 function needsPermission(permission) {
     return function(req,res,next) {
         if (settings && settings.adminAuth) {
-            return passport.authenticate(['bearer','anon'],{ session: false })(req,res,function() {
+            return passport.authenticate(['bearer','anon','tokens'],{ session: false })(req,res,function() {
```
[the remark this answers is not in the document]
Comment (KazuhiroItoh) [reply packages/node_modules/@node-red/editor-api/lib/auth/index.js]: Fixed "tokens" order. [458d794](https://github.com/node-red/node-red/pull/2479/commits/458d794f52492a311316add9486a3986c2b64904)

File: packages/node_modules/@node-red/editor-api/lib/auth/strategies.js [hunk 2]
```diff
@@ -123,9 +123,31 @@ AnonymousStrategy.prototype.authenticate = function(req) {
...
 }
 
+function TokensStrategy() {
+  passport.Strategy.call(this);
+  this.name = 'tokens';
+}
+util.inherits(TokensStrategy, passport.Strategy);
+TokensStrategy.prototype.authenticate = function(req) {
+    var self = this;
+    var token = req.headers[Users.tokenHeader()];
```
[the remark this answers is not in the document]
Comment (KazuhiroItoh) [reply packages/node_modules/@node-red/editor-api/lib/auth/strategies.js]: Fixed to pass the actual token to the plugin. [83942c2](https://github.com/node-red/node-red/pull/2479/commits/83942c255151aee0e92b2045b93a927576483f54)

**Comment by knolleary** [comment]
Thank you for making those updates. Looks good to me.

As a separate issue, I think there is another requirement for adminAuth that is related to this. We have been asked whether it is possible to authenticate a request based on its IP address or other features of the request itself.

I am not suggesting we address that requirement with this PR - but we may end up adding another auth option that exposes the entire request object to allow for more detailed auth schemes. That option could also be used to achieve the same thing as this one. But on balance I think it is helpful to provide this implementation as well.

**Comment by eapenjohn** [comment]
i dont see any release tag associated with this issue, Do you able to release this ?

**Comment by knolleary** [comment]
@eapenjohn it will be in the 1.1.0 release.

**Comment by eapenjohn** [comment]
Do you have any time line for the 1.1.0 release?

**Comment by knolleary** [comment]
Should have an initial beta out in the next week or so.

**Comment by eapenjohn** [comment]
@knolleary thanks

**Pull request merged**